Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -344,6 +344,26 @@ private void StartWorkers(LoggingContext loggingContext)
             m_allWorker = new AllWorker(m_workers.ToArray());
 
             ExecutionLog?.WorkerList(new WorkerListEventData { Workers = m_workers.SelectArray(w => w.Name) });
+        }
+
+        private readonly TaskSourceSlim<bool> m_schedulerCompletionExceptMaterializeOutputs = TaskSourceSlim.Create<bool>();
+        private bool AnyPendingPipsExceptMaterializeOutputs()
+        {
+            long numRunningOrQueued = m_pipQueue.NumRunningOrQueued;
+            long numRunningOrQueuedExceptMaterialize = numRunningOrQueued - m_pipQueue.GetNumRunningByKind(DispatcherKind.Materialize) - m_pipQueue.GetNumQueuedByKind(DispatcherKind.Materialize);
+
+            if (numRunningOrQueuedExceptMaterialize == 0)
+            {
+                RetrievePipStateCounts(out long totalPips, out long readyPips, out long waitingPips, out long runningPips, out long donePips, out long failedPips, out long skippedPips, out long ignoredPips);

```

Comment:
>RetrievePipStateCounts(out long totalPips, out long readyPips, out long waitingPips, out long runningPips, out long donePips, out long failedPips, out long skippedPips, out long ignoredPips); [](start = 16, length = 191)

why do we need this? isn't `m_pipQueue.NumRunningOrQueued` already capturing all the pips?

Reply:
Added comment.

---
In reply to: [348746468](https://github.com/microsoft/BuildXL/pull/1216#discussion_r348746468) [](ancestors = 348746468)